swaps/swaps_service.go

```diff
@@ -1317,3 +1346,134 @@ func (svc *swapsService) doMempoolRequest(endpoint string, result interface{}) e
 	}
 	return nil
 }
+
+func (svc *swapsService) deriveAddressFromXpub(xpub string, index uint32) (string, error) {
+	var netParams *chaincfg.Params
+	switch svc.cfg.GetNetwork() {
+	case "bitcoin", "mainnet":
+		netParams = &chaincfg.MainNetParams
+	case "testnet":
+		netParams = &chaincfg.TestNet3Params
+	case "regtest":
+		netParams = &chaincfg.RegressionNetParams
+	case "signet":
+		netParams = &chaincfg.SigNetParams
+	default:
+		return "", fmt.Errorf("unsupported network: %s", svc.cfg.GetNetwork())
+	}
+
+	extPubKey, err := hdkeychain.NewKeyFromString(xpub)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse xpub: %w", err)
+	}
+
+	externalChain, err := extPubKey.Derive(0)
+	if err != nil {
+		return "", fmt.Errorf("failed to derive external chain: %w", err)
+	}
+
+	addressKey, err := externalChain.Derive(index)
+	if err != nil {
+		return "", fmt.Errorf("failed to derive address key at index %d: %w", index, err)
+	}
+
+	pubKey, err := addressKey.ECPubKey()
+	if err != nil {
+		return "", fmt.Errorf("failed to get public key: %w", err)
+	}
+
+	pubKeyHash := btcutil.Hash160(pubKey.SerializeCompressed())
+	address, err := btcutil.NewAddressWitnessPubKeyHash(pubKeyHash, netParams)
+	if err != nil {
+		return "", fmt.Errorf("failed to create address: %w", err)
+	}
+
+	return address.EncodeAddress(), nil
+}
+
+func (svc *swapsService) checkAddressHasTransactions(address string, esploraApiRequester func(endpoint string) (interface{}, error)) (bool, error) {
+	response, err := esploraApiRequester("/address/" + address + "/txs")
+	if err != nil {
+		return false, fmt.Errorf("failed to get address transactions: %w", err)
+	}
+
+	transactions, ok := response.([]interface{})
+	if !ok {
+		return false, fmt.Errorf("unexpected response format from esplora API")
+	}
+
+	return len(transactions) > 0, nil
+}
+
+func (svc *swapsService) getNextUnusedAddressFromXpub() (string, uint32, error) {
+	destination, _ := svc.cfg.Get(config.AutoSwapDestinationKey, "")
+	if destination == "" {
+		return "", 0, errors.New("no destination configured")
+	}
+
+	if err := svc.validateXpub(destination); err != nil {
+		return "", 0, errors.New("destination is not a valid XPUB")
+	}
+
+	indexStr, _ := svc.cfg.Get(config.AutoSwapXpubIndexStart, "0")
```

Comment:
the 2nd parameter is the encryption key, this should be empty. I will fix it in a following commit